src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,279 @@
+        }
+
+        /// <inheritdoc/>
+        public override ComponentInformation QueryComponentInformation()
+        {
+            ComponentInformation self = new ComponentInformation(this, "Virtual GPIO Controller");
+
+            self.Properties["OpenPins"] = string.Join(", ", _gpioPins.Select(x => x.Key));
+
+            return self;
+        }
+
+        /// <inheritdoc/>
+        public override PinValue Read(int pinNumber)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not read from pin {pinNumber} because it is not open.");
+            }
+
+            return _pins[pinNumber].Read();
+        }
+
+        /// <summary>
+        /// Sets the mode to a pin.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="mode">The mode to be set.</param>
+        public override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not set a mode to pin {pinNumber} because it is not open.");
+            }
+
+            if (!IsPinModeSupported(pinNumber, mode))
+            {
+                throw new InvalidOperationException($"Pin {pinNumber} does not support mode {mode}.");
+            }
+
+            _pins[pinNumber].SetPinMode(mode);
+        }
+
+        /// <inheritdoc/>
+        public override void Write(int pinNumber, PinValue value)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not write to pin {pinNumber} because it is not open.");
+            }
+
+            _pins[pinNumber].Write(value);
+        }
+
+        /// <inheritdoc/>
+        public override void Toggle(int pinNumber)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not read from pin {pinNumber} because it is not open.");
+            }
+
+            _pins[pinNumber].Toggle();
+        }
+
+        /// <inheritdoc/>
+        public override void RegisterCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventTypes, PinChangeEventHandler callback)
+        {
+            if (_callbackEvents.Count == 0)
+            {
+                _pins[pinNumber].ValueChanged += PinValueChanged;
+            }
+
+            _callbackEvents.TryAdd(pinNumber, new PinEvent(eventTypes, callback));
+        }
+
+        private void PinValueChanged(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
+        {
+            // Get GpioPin associated with the real life pin
+            // Note that if we have multiple pins with the same pin number, we will only take the first one in the list
+            var pins = _pins.Where(m => m.Value.PinNumber == pinValueChangedEventArgs.PinNumber);
```

Comment:
You could possibly check whether the sender is equal to the controller of the pin to solve this.

Reply:
I would love to but the original controller is not available. Only the GpioPin is available.